core/utils/utils.go

```diff
@@ -515,3 +515,16 @@ func Uint256ToHex(n *big.Int) (string, error) {
 	}
 	return common.BigToHash(n).Hex(), nil
 }
+
+// SafeByteSlice returns an error on out of bounds access to a byte array, where a
+// normal slice would panic instead
+func SafeByteSlice(ary []byte, start int, end int) (slice []byte, err error) {
+	defer func() {
```

Comment:
I'd prefer a different solution where we don't use recover, but instead do a bounds check.

Reply:
Strong agree on this one.

Reply:
Out of interest, why?

Reply:
I originally had a bounds check version but there were a lot of different options (e.g. start greater than end, negative indexes) and it felt safer just to recover from panic.

Reply:
A defer will rescue any kind of panic, some of which may not be related to your failure, the code should be simpler `if x > len(arr) { return err }` right? Also panics cause a bit of a performance impact, due to stack unwinding and adding handlers to the stack for every invocation of he defer.

Reply:
Not that simple.

What if start is negative? What if start > end? There are a bunch of different things that can go wrong here. It feels safer to simply recover.